### DISCUSSION TITLE
v2.21

### DISCUSSION THREAD

**Discussion by Ralim** [Ralim/IronOS#1660] [General]
## Key changes

* Bluetooth Low Energy support for PinecilV2
* Large cleanup of translation files; and refactor of how we handle fonts for translations
* Fixes for I2C corruption on PinecilV2
* Option for using adjustable profiles on USB-PD or not
* Cleanups and improvements to the generated [documents website](https://ralim.github.io/IronOS/)

For PinecilV2 users [blisp](https://github.com/pine64/blisp/releases) is currently my reccomended CLI tool for updating the device. It is built for all main OS's automatically.

## What's Changed
* Update translation_HU.json by @lalamper in https://github.com/Ralim/IronOS/pull/1505
* Update translation_IT.json by @WillCap397 in https://github.com/Ralim/IronOS/pull/1504
* Updated translation_HR.json by @jonnieZG in https://github.com/Ralim/IronOS/pull/1499
* Update translation_RU.json by @JugglerLKR in https://github.com/Ralim/IronOS/pull/1500
* removed some unused lines, ... by @discip in https://github.com/Ralim/IronOS/pull/1502
* makes the 'boost mode' deactivatable (as originally intended) by @discip in https://github.com/Ralim/IronOS/pull/1507
* update translation by @dmitrygribenchuk in https://github.com/Ralim/IronOS/pull/1509
* update translation by @dmitrygribenchuk in https://github.com/Ralim/IronOS/pull/1511
* fixes cooldown flashing by @discip in https://github.com/Ralim/IronOS/pull/1515
* updated Menu.md by @discip in https://github.com/Ralim/IronOS/pull/1512
* Reboot after reset by @discip in https://github.com/Ralim/IronOS/pull/1513
* Fix the calibrate voltage screen, when it exits, it used to overlay t… by @purdeaandrei in https://github.com/Ralim/IronOS/pull/1530
* Miniware+Pinecil: fix number of items in uVtoDegC array. This only af… by @purdeaandrei in https://github.com/Ralim/IronOS/pull/1531
* Expanded the build script to build customizable multiple languages and models firmwares by @GeminiServer in https://github.com/Ralim/IronOS/pull/1520
* New translations format by @Ralim in https://github.com/Ralim/IronOS/pull/1493
* Just a small typo correction and a suggestion for clearer wording. by @discip in https://github.com/Ralim/IronOS/pull/1533
* Ble by @Ralim in https://github.com/Ralim/IronOS/pull/1434
* Settings toggle for ble by @Ralim in https://github.com/Ralim/IronOS/pull/1537
* Update translation_IT.json by @federicodilo in https://github.com/Ralim/IronOS/pull/1539
* Merging newest translations into dev by @github-actions in https://github.com/Ralim/IronOS/pull/1540
* Drop decimal when > 99.9W by @Ralim in https://github.com/Ralim/IronOS/pull/1536
* Merging newest translations into dev by @github-actions in https://github.com/Ralim/IronOS/pull/1542
* simplification by @discip in https://github.com/Ralim/IronOS/pull/1543
* fixing languages & reverting unwanted weblate changes by @discip in https://github.com/Ralim/IronOS/pull/1548
* Change timing scheduler to only run main timer once ADC is done by @Ralim in https://github.com/Ralim/IronOS/pull/1535
* track and return Operating Mode with BLE by @TomW1605 in https://github.com/Ralim/IronOS/pull/1553
* expand bulk data to match individual value data by @TomW1605 in https://github.com/Ralim/IronOS/pull/1555
* Update translation_FR.json by @Alexandrebafoil in https://github.com/Ralim/IronOS/pull/1556
* Use the scan response to advertise some of the available service UUID… by @cohaolain in https://github.com/Ralim/IronOS/pull/1557
* Small update in translation_IT.json by @federicodilo in https://github.com/Ralim/IronOS/pull/1565
* BLE Settings rename + Unique Name by @Ralim in https://github.com/Ralim/IronOS/pull/1567
* Fixed USB-PD policy bug. by @Samuelrmlink in https://github.com/Ralim/IronOS/pull/1572
* Added VPDO option for disabling PPS/EPR PDOs. by @Samuelrmlink in https://github.com/Ralim/IronOS/pull/1582
* Update README.md to explain why TS80 is not recommended by @DjordjeMandic in https://github.com/Ralim/IronOS/pull/1588
* Fix updating certain settings over BLE by @NeilHanlon in https://github.com/Ralim/IronOS/pull/1592
* minor formatting fix by @discip in https://github.com/Ralim/IronOS/pull/1593
* update translation by @dmitrygribenchuk in https://github.com/Ralim/IronOS/pull/1596
* Update documents, esp. Pinecil V2 instructions by @River-Mochi in https://github.com/Ralim/IronOS/pull/1604
* Fix README badges by @FoxieDev in https://github.com/Ralim/IronOS/pull/1607
* Update README.md to fix small typo by @DjordjeMandic in https://github.com/Ralim/IronOS/pull/1611
* Updated translation_DE.json by @StevenSeifried in https://github.com/Ralim/IronOS/pull/1617
* Add new EL translation strings by @Devnol in https://github.com/Ralim/IronOS/pull/1623
* update translation_pt.jason by @ViPeR5000 in https://github.com/Ralim/IronOS/pull/1612
* Update configuration.h by @River-Mochi in https://github.com/Ralim/IronOS/pull/1625
* Update translation_EN.json by @discip in https://github.com/Ralim/IronOS/pull/1626
* Update translation_PT.json by @G-Pereira in https://github.com/Ralim/IronOS/pull/1627
* Update translation_RO.json by @mobycl1ck in https://github.com/Ralim/IronOS/pull/1628
* Update settings icon description by @patridge in https://github.com/Ralim/IronOS/pull/1631
* Add FW version to BLE characteristic by @ew1abz in https://github.com/Ralim/IronOS/pull/1633
* Add building steps for Pinecil V2 to Development.md by @ew1abz in https://github.com/Ralim/IronOS/pull/1634
* adapting 10 °C > 50 °F by @discip in https://github.com/Ralim/IronOS/pull/1636
* Readjusting brightness (Pinecil v2) by @discip in https://github.com/Ralim/IronOS/pull/1622
* Update translation_RO.json by @discip in https://github.com/Ralim/IronOS/pull/1632
* Update translation_DE.json by @StevenSeifried in https://github.com/Ralim/IronOS/pull/1637
* Change to 128bit UUID's by @Ralim in https://github.com/Ralim/IronOS/pull/1608
* Added Variables to Dockerfile by @slycordinator in https://github.com/Ralim/IronOS/pull/1597
* implementing #1594 by @discip in https://github.com/Ralim/IronOS/pull/1616
* USBPDDebug: Check whether VPDO is EPR or PPS by @Samuelrmlink in https://github.com/Ralim/IronOS/pull/1600
* fixing power limit display by @discip in https://github.com/Ralim/IronOS/pull/1641
* fix Mkdocs edit button by @River-Mochi in https://github.com/Ralim/IronOS/pull/1642
* fix mkdocs  markdown extensions by @River-Mochi in https://github.com/Ralim/IronOS/pull/1648
* add plugins for mkdocs features by @River-Mochi in https://github.com/Ralim/IronOS/pull/1650
* fix typo on run command alignment by @River-Mochi in https://github.com/Ralim/IronOS/pull/1652
* Update translation_IT.json by @federicodilo in https://github.com/Ralim/IronOS/pull/1654
* correcting MAX_POWER_LIMIT to 120 by @discip in https://github.com/Ralim/IronOS/pull/1658
* unifying by @discip in https://github.com/Ralim/IronOS/pull/1657

## New Contributors
* @lalamper made their first contribution in https://github.com/Ralim/IronOS/pull/1505
* @WillCap397 made their first contribution in https://github.com/Ralim/IronOS/pull/1504
* @purdeaandrei made their first contribution in https://github.com/Ralim/IronOS/pull/1530
* @TomW1605 made their first contribution in https://github.com/Ralim/IronOS/pull/1553
* @Alexandrebafoil made their first contribution in https://github.com/Ralim/IronOS/pull/1556
* @cohaolain made their first contribution in https://github.com/Ralim/IronOS/pull/1557
* @Samuelrmlink made their first contribution in https://github.com/Ralim/IronOS/pull/1572
* @DjordjeMandic made their first contribution in https://github.com/Ralim/IronOS/pull/1588
* @NeilHanlon made their first contribution in https://github.com/Ralim/IronOS/pull/1592
* @FoxieDev made their first contribution in https://github.com/Ralim/IronOS/pull/1607
* @StevenSeifried made their first contribution in https://github.com/Ralim/IronOS/pull/1617
* @ViPeR5000 made their first contribution in https://github.com/Ralim/IronOS/pull/1612
* @G-Pereira made their first contribution in https://github.com/Ralim/IronOS/pull/1627
* @mobycl1ck made their first contribution in https://github.com/Ralim/IronOS/pull/1628
* @patridge made their first contribution in https://github.com/Ralim/IronOS/pull/1631
* @ew1abz made their first contribution in https://github.com/Ralim/IronOS/pull/1633

**Full Changelog**: https://github.com/Ralim/IronOS/compare/v2.20...v2.21

<hr /><em>This discussion was created from the release <a href='https://github.com/Ralim/IronOS/releases/tag/v2.21'>v2.21</a>.</em>

**Comment by khjkuo** [comment]
dear sir:
I have a problem that has troubled me for a long time. I have a ts100 electric soldering iron with built-in GD main chip DFU3.48. I want to upgrade the third-party firmware RALIM/IRONOS TS100V2.21 firmware cannot be upgraded
Please help me as an attachment, thank you
E-mail: khj.kuo@gmail.com

![ts100 ng](https://github.com/Ralim/IronOS/assets/77198229/c5634626-826a-4f55-becd-8f74aa158a1c)

**Reply by ia** [reply]
Hello. Let me try to help you and clarify a few things.

> I have a problem that has troubled me for a long time.

First of all, [here is the direct link to the form](https://github.com/Ralim/IronOS/issues/new/choose) which should be filled by a user who have any particular problem/issue/bug/etc with the software/hardware related to `IronOS` project. Please, try to use that form in the next time for any issue. And try to include as much details as possible - the more information you provide, the more chance that others will be able to figure out what could possibly be wrong.

> E-mail:

Second, `IronOS` is [free software](https://www.gnu.org/philosophy/free-sw.html) project so interactions are endorsed to be open & transparent so the whole community could benefit from issues solved by other users for other users. Users shouldn't be worry nor ashamed to discuss their bugs with the community using [public official issue tracker](https://github.com/Ralim/IronOS/issues) regardless the complexity of the issue.

> attachment

Third, just like any big community of any free software project, `IronOS` has worldwide community from all over the world, literally - from different countries with different languages & cultures. However, English is a _default language_ for projects like this - in such way there are more chances that more people will be able to help each other. So, could you, somehow, copy-paste the text of the error message from that message box into your report, or at least briefly provide translation of what it says exactly? I have non-English interface myself on some machines but providing screenshots from them for bug report will be unclear for others as well.

And, fourth, regardless the nature of your problem, here is some recommendations which usually help to most of the users with their problems updating & flashing their `Miniware` `TS` models since built-in `DFU` from `Miniware` is not quite stable & reliable indeed, _that is well-known issue_. So try these recommendations:

- rename file which you try to flash from `TS100_ZH_TW.hex`  to `FW.HEX` (preserving the capital case) - `DFU` may be very sensitive to case sometimes
- unplug `TS100` from `PC`
- press & hold **`A`** button (the closest to the tip)
- plug cable from `TS100` to `PC`
- make sure that you see the line on the screen: "`DFU3.48`"
- make sure that `removable USB drive` appeared
- drag'n'drop `FW.HEX` to `USB drive`
- wait ~30 seconds at least
- "safely eject" `USB drive`

Now, if you still have the issue copying the file, then try these recommendations - they usually help as well to other users:

- **do not use USB hub** while connecting to PC and **do not use front USB panel** (if it's "tower"-like PC);
- try different cables / ports on your PC while plugging-in
- try to connect `TS100` in `DFU` mode to other PC(s) with other OS(es) if possible (at least just to check if the issue appears regardless OS / computer)

Maybe this help, maybe not. But I hope at least something from these advises will work for you. Good luck.

**Reply by khjkuo** [reply]
dear ia:
First of all, thank you for your detailed explanation. This is the first time I have reported a problem. I don’t understand some rules, so please forgive me for any mistakes. I come from Taiwan and use a laptop with Chinese fonts, so the pictures cannot be translated into English. Please forgive me. Miniware’s official explanation is:" DFU is not open source, it needs Ralim/Iron firmware to be compatible with DFU3.48. "**_Can I ask Ralim to release a new version of firmware to be compatible with DFU3.48?_** Thank you very much.

**Reply by Ralim** [reply]
The firmware _is_ compatible to DFU 3.48.
One of my test Irons runs it and I test it before release.
However DFU 3.48 is _not_ stable. Generally to get it to load you need to do the double copy method as detailed elsewhere on the docs. And even then some USB controllers the GD clone chips just dont work well one as Miniware didnt fit a crystal oscillator as required by the spec.

**Reply by khjkuo** [reply]
Dear Ralim:
"Generally to get it to load you need to do the double copy method as detailed elsewhere on the docs" I don't understand what you mean, please explain in detail or if you have a case, you can show me or link the URL for my reference Learn to solve thanks a lot

**Reply by khjkuo** [reply]
Dear Ralim:
What is the use of IronOS-DFU? Since I can’t use DFU3.48 firmware, can I use IronOS-DFU to solve the upgrade problem? Do you have a video for my reference or do you have a sample link for my reference? Thank you very much

**Reply by ia** [reply]
> I don’t understand some rules, so please forgive me for any mistakes.

Please, you don't have to apologize for anything - we all learn something new everyday. I just provided you a few advises so you could get the better result in getting help, that's it.

> the pictures cannot be translated into English.

What I meant is that _it maybe could be useful_ if you could provide in English briefly the short translation of what exactly that error message box says. It could be helpful or not. But the more information you provide when you report the issue, the better.

> you can show me or link the URL for my reference

- open this [link](https://ralim.github.io/IronOS/Flashing/TS100) with official documentation
- scroll down to **`FAQ`** section
- read subsection with title **`The file is showing up with the extension .ERR`**

It _may be_ your case. **However**, it seems your `Windows` can't even copy the firmware file to the `DFU USB drive` (which your error message box says on your screenshot but I'm not sure - that's why it would be useful to understand what it says exactly).

Anyway, did you try to follow common recommendations from [my comment](https://github.com/Ralim/IronOS/discussions/1660#discussioncomment-6906193)? These are common recommendations which _usually_ help other users with the similar issues like the one that you have now. And it's based on the advises which did help others (including myself) to solve the similar issues. Did you rename the firmware file to very short name with all capital letters? Did you even try to change the cable? What are the results? The same? Or the error message looked _different_? Do you have `ERR` file on `DFU USB drive` after copying firmware or you didn't? Did you try different USB port on your computer (remember - no hub and no front panel, no any usb extension cords)? Was there any difference? Did you try to use another PC? Usually these are the ones things which help in your case.

**Comment by khjkuo** [comment]
Dear ia:
First of all, thank you for explaining that I have tried all your methods. I still can’t upgrade the IronOS firmware, but if I upgrade the official firmware 2.20v, it will succeed. How strange? What’s the problem?
My laptop usb3.1 interface is connected to ts100 to copy the IronOS firmware and paste the virtual disk DFU screen appears
Error 0x800701B1: A device that does not exist was specified

**Reply by ia** [reply]
> I have tried all your methods

So, did you try to copy the firmware with short name using different USB cable with different PC with different Windows version? And even on a different PC the issue is exactly the same?

**Reply by khjkuo** [reply]
Dear ia:
I have tried all the methods you taught me, but I still can't upgrade the IronOS firmware. I have also tried different computers and different USB cables, but I still can't upgrade the IronOS firmware. Can you please @Ralim to help me Thank you very much

**Reply by Ralim** [reply]
@khjkuo 
There is no reason to keep tagging me.
@ia is telling you everything I would as well.
The firmware has been tested on TS100,TS80,TS80P,TS101 units here and flashes and works fine for me.
I am however using linux rather than Windows. Its been well documented that newer Windows versions have issues with the bootloader that comes on the TS100. They never acted on my reports of issues or suggestions for fixes, and instead just always say the same story that its my end. Yet it works perfectly for me, and if you read the existing issue and discussions the fixes tend to be to try using a different computer / os. This is because they have coded their bootloader with assumptions around how the OS acts, and the exact order it does things in. Any anti-virus software for example and interfere with this. Which is why we suggest using a different PC; as the hope is that one of them is "clean" of interfering software.

**Reply by khjkuo** [reply]
Dear Ralim&ia:
Once again, I feel your warm service and teaching me. Although my problem has not been solved yet, I believe that people who work hard will find a way to solve it. Finally, thank you very much
